### PR TITLE
TW-786 Refactored transaction history for send, receive and delegate

### PR DESCRIPTION
**Pull request by herkoss**

### CODE REVIEW

File: src/components/activity-groups-list/activity-group-item/activity-group-item.tsx [hunk 1]
```diff
@@ -25,45 +31,75 @@ interface Props {
...
-      <View style={styles.lowerContainer}>
-        <View style={styles.statusContainer}>
-          <ActivityStatusBadge status={firstActivity.status} />
+      <Divider size={formatSize(12)} />
+      <View style={[styles.row, styles.justifyBetween]}>
+        <View style={styles.row}>
+          <ActivityStatusBadge status={group[0].status} />
           <Divider size={formatSize(4)} />
-          <ActivityTime timestamp={firstActivity.timestamp} />
+          <ActivityTime timestamp={1689343447} />
```
Comment (lendihop) [inline src/components/activity-groups-list/activity-group-item/activity-group-item.tsx:69]: const FIRST_ACTIVITY_TIMESTAMP = 1689343447;

File: src/components/activity-groups-list/activity-groups-list.tsx [hunk 2]
```diff
@@ -54,49 +54,55 @@ export const ActivityGroupsList: FC<Props> = ({
     return result;
   }, [activityGroups]);
 
+  console.log('sections: ', sections[0]?.title, JSON.stringify(sections[0]?.data[0], null, 2));
+
```
Comment (lendihop) [inline src/components/activity-groups-list/activity-groups-list.tsx:58]: rm logs

File: src/components/activity-groups-list/activity-group-item/activity-group-item.tsx (hunk 1 above)
Comment (keshan3262) [reply src/components/activity-groups-list/activity-group-item/activity-group-item.tsx:69]: Looks like a mock value... Could you replace it with actual one?

File: src/components/activity-groups-list/activity-group-item/activity-group-type/use-activity-group-info.hook.ts [hunk 3]
```diff
@@ -1,47 +1,75 @@
...
+import { useSelectedAccountSelector } from 'src/store/wallet/wallet-selectors';
+import { useColors } from 'src/styles/use-colors';
+import { truncateLongAddress } from 'src/utils/exolix.util';
+import { isString } from 'src/utils/is-string';
+
+interface ReturnType {
+  transactionType: string;
+  transactionSubtype: string;
+  transactionHash: string;
+  destination: [string, string, string];
```
Comment (keshan3262) [inline src/components/activity-groups-list/activity-group-item/activity-group-type/use-activity-group-info.hook.ts:14]: Sorry, this type looks non-informative. Could you use an object instead?

**Review by keshan3262 — DISMISSED** [review]

**Review by lendihop — DISMISSED** [review]

File: ios/TempleWallet.xcodeproj/project.pbxproj
Comment (lendihop) [inline ios/TempleWallet.xcodeproj/project.pbxproj:1]: pls revert this changes

File: src/components/activity-groups-list/activity-group-item/activity-group-details.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      {areDetailsVisible && (
+        <View style={styles.card}>
+          {!isEmpty(nonZeroAmounts.amounts) && (
+            <View style={[styles.detailItem, styles.detailItemBorder, styles.itemsStart]}>
+              <Text style={styles.detailText}>{transactionSubtype}</Text>
+              <View style={styles.row}>
+                <View style={styles.itemsEnd}>
+                  <ActivityGroupAmountChange nonZeroAmounts={nonZeroAmounts} textSize={TextSize.Small} />
+                  {isCollectible ? (
+                    <Text style={styles.ntfPrice}>0.00$</Text>
```
Comment (keshan3262) [inline src/components/activity-groups-list/activity-group-item/activity-group-details.tsx:78]: Please use `useCurrentFiatCurrencyMetadataSelector` to get the actual symbol to be used

**Review by lendihop — APPROVED** [review]

**Review by keshan3262 — DISMISSED** [review]

**Review by lendihop — APPROVED** [review]

File: src/components/activity-groups-list/activity-group-item/activity-group-type/use-activity-group-info.hook.ts [hunk 5]
```diff
@@ -11,14 +12,18 @@ interface ActivityInfo {
...
     label: string;
     value: string;
     address: string | undefined;
+    bakerLogo?: string;
   };
 }
 
 export const BAKING_REWARDS_TEXT = 'Baking rewards';
+export const DELEGATION_TEXT = 'Delegation';
+export const UNDELEGATION_TEXT = 'Undelegation';
```
Comment (lendihop) [inline src/components/activity-groups-list/activity-group-item/activity-group-type/use-activity-group-info.hook.ts]: mb export this as a enum TransactionTypes?

**Review by keshan3262 — APPROVED** [review]

**Review by lendihop — APPROVED** [review]

File: src/components/activity-groups-list/activity-group-item/activity-details-card/receive-tokens-details.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        <View>
+          <ActivityGroupAmountChange nonZeroAmounts={nonZeroAmounts} textSize={TextSize.Small} />
+          <ActivityGroupDollarAmountChange nonZeroAmounts={nonZeroAmounts} />
+        </View>
+      </View>
+      <View style={[styles.itemWrapper, styles.border]}>
+        <Text style={styles.text}>From:</Text>
+        <WalletAddress isLocalDomainNameShowing publicKeyHash={address} />
+      </View>
+      <View style={styles.itemWrapper}>
```
Comment (keshan3262) [inline src/components/activity-groups-list/activity-group-item/activity-details-card/receive-tokens-details.tsx:40]: Sorry, could you create components for common rows like this one? I think that it won't break compatibility with higher versions of the transactions parser

File: src/interfaces/activity.interface.ts [hunk 7]
```diff
@@ -1,31 +1,17 @@
...
 
-export const emptyActivity: ActivityInterface = {
-  type: ActivityTypeEnum.Transaction,
-  status: ActivityStatusEnum.Pending,
+export const emptyActivity: Activity = {
+  type: ActivityType.Send,
+  status: TzktOperationStatus.Applied,
   hash: '',
-  amount: '',
+  tokensDeltas: [{ atomicAmount: new BigNumber(0), tokenSlug: 'tez' }],
```
Comment (keshan3262) [inline src/interfaces/activity.interface.ts:12]: Please use `TEZ_TOKEN_SLUG` here

**Review by keshan3262 — DISMISSED** [review]

File: src/hooks/use-contract-activity.ts [hunk 8]
```diff
@@ -1,25 +1,48 @@
...
 
-import { ActivityGroup } from '../interfaces/activity.interface';
 import { UseActivityInterface } from '../interfaces/use-activity.interface';
 import { useSelectedRpcUrlSelector } from '../store/settings/settings-selectors';
 import { useSelectedAccountSelector } from '../store/wallet/wallet-selectors';
 import { isDefined } from '../utils/is-defined';
 import { loadActivity } from '../utils/token-operations.util';
 
+const KNOWN_BAKERS: Array<TzktMemberInterface> = [
+  { address: EVERSTAKE_PAYOUTS_BAKER.address, alias: 'Everstake payouts' }
```
Comment (keshan3262) [inline src/hooks/use-contract-activity.ts:18]: Sorry, I have discovered plenty of payouts addresses recently:
```
[
  {
    "address": "tz1SvASb4xmkiM82ijexDHzn6bYakaReBypt",
    "name": "Anonstake Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1SvASb4xmkiM82ijexDHzn6bYakaReBypt"
  },
  {
    "address": "tz1TqU5hwh6CtC8Ps5GYBi7YyawvtKLNCbvQ",
    "name": "Ateza Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1TqU5hwh6CtC8Ps5GYBi7YyawvtKLNCbvQ"
  },
  {
    "address": "tz1Zrqm4TkJwqTxm5TiyVFh6taXG4Wrq7tko",
    "name": "Bake’n’Rolls Payouts 2",
    "logo": null
  },
  {
    "address": "tz1iVCqjMRG7MfBoFnVjbLQyeLAeJQfLxXze",
    "name": "Bakery-IL Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1iVCqjMRG7MfBoFnVjbLQyeLAeJQfLxXze"
  },
  {
    "address": "tz1NAozDvi5e7frVq9cUaC3uXQQannemB8Jw",
    "name": "BakeTz Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1NAozDvi5e7frVq9cUaC3uXQQannemB8Jw"
  },
  {
    "address": "tz1PayLDWLg39fbkwKejnT3aijz4QEd5kgt3",
    "name": "Baking Team Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1PayLDWLg39fbkwKejnT3aijz4QEd5kgt3"
  },
  {
    "address": "tz1UQvdfQ2cwyDa6k6U62mkHWeULwBqCcJwN",
    "name": "Citadel.one Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1UQvdfQ2cwyDa6k6U62mkHWeULwBqCcJwN"
  },
  {
    "address": "tz1cNARmnRRrvZgspPr2rSTUWq5xtGTuKuHY",
    "name": "Cryptium Labs Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1cNARmnRRrvZgspPr2rSTUWq5xtGTuKuHY"
  },
  {
    "address": "tz1ZbSrRrfhU8LYHELWNswx2JcFARXTGKKVk",
    "name": "DNAG Delegation Payouts 1",
    "logo": "https://services.tzkt.io/v1/avatars/tz1ZbSrRrfhU8LYHELWNswx2JcFARXTGKKVk"
  },
  {
    "address": "tz1aqYbUuXoraB7h39Ls3HvbDwiMhHRV199L",
    "name": "Elite Tezos Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1aqYbUuXoraB7h39Ls3HvbDwiMhHRV199L"
  },
  {
    "address": "tz1W1en9UpMCH4ZJL8wQCh8JDKCZARyVx2co",
    "name": "Everstake Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1W1en9UpMCH4ZJL8wQCh8JDKCZARyVx2co"
  },
  {
    "address": "tz1dXCHSN8uwnappk4uet1WQeuvbaJog4Jwi",
    "name": "Figment Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1dXCHSN8uwnappk4uet1WQeuvbaJog4Jwi"
  },
  {
    "address": "tz1ePtw5wtQMNH1LkJaVXbeVmW7joWqdNzge",
    "name": "GOLD TZ Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1ePtw5wtQMNH1LkJaVXbeVmW7joWqdNzge"
  },
  {
    "address": "tz1TyAEPtLt7oapQvZirVbswz1VRbdRqdfW7",
    "name": "Happy Tezos Dexter Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1TyAEPtLt7oapQvZirVbswz1VRbdRqdfW7"
  },
  {
    "address": "tz1hodLFy8YDHtf1qJ3XH1gCyeXfwWEitkj9",
    "name": "hodl.farm Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1hodLFy8YDHtf1qJ3XH1gCyeXfwWEitkj9"
  },
  {
    "address": "tz1MTAmuFmKu3MoxnkVsMj7xvSScj6d83hu5",
    "name": "LetzBake! Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1MTAmuFmKu3MoxnkVsMj7xvSScj6d83hu5"
  },
  {
    "address": "tz1fRAiiDUxnXZwDmQghh1n6VytQUZWNPwWN",
    "name": "Melange Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1fRAiiDUxnXZwDmQghh1n6VytQUZWNPwWN"
  },
  {
    "address": "tz1MdSsDBG8k8q1pxZrdpSC2Aw49ukFs8ACy",
    "name": "Money Every 3 Days Payouts 1",
    "logo": "https://services.tzkt.io/v1/avatars/tz1MdSsDBG8k8q1pxZrdpSC2Aw49ukFs8ACy"
  },
  {
    "address": "tz1XVBuWfdFpWCiMvsEV8KEX2nkMHN3mfuHJ",
    "name": "Neokta Labs Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1XVBuWfdFpWCiMvsEV8KEX2nkMHN3mfuHJ"
  },
  {
    "address": "tz1c9HEyZdAWXZCLE7h3KcZCyykjZr7VY3QM",
    "name": "PayTezos Payouts 2",
    "logo": "https://services.tzkt.io/v1/avatars/tz1c9HEyZdAWXZCLE7h3KcZCyykjZr7VY3QM"
  },
  {
    "address": "tz1MoonPbyMJSqMVsVwExgVc5egnv18CgSDq",
    "name": "P2P Validator Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1MoonPbyMJSqMVsVwExgVc5egnv18CgSDq"
  },
  {
    "address": "tz1VCP3MQdfxtRc83CRJZX43wc1oZUxb6KmH",
    "name": "Spice Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1VCP3MQdfxtRc83CRJZX43wc1oZUxb6KmH"
  },
  {
    "address": "tz1fishChEERnSesm8mHe7xsJJwymbHVLcYf",
    "name": "Stake.fish Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1fishChEERnSesm8mHe7xsJJwymbHVLcYf"
  },
  {
    "address": "tz1Ywgcavxq9D6hL32Q2AQWHAux9MrWqGoZC",
    "name": "StakeNow Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1Ywgcavxq9D6hL32Q2AQWHAux9MrWqGoZC"
  },
  {
    "address": "tz1Y2rhY9nBKfd2jp7m4z8iAMKKARWS3BvFA",
    "name": "Shake 'n Bake Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1Y2rhY9nBKfd2jp7m4z8iAMKKARWS3BvFA"
  },
  {
    "address": "tz1gs78wrxS34XhGPECmN5fABKpEuhUGwAXS",
    "name": "Stakery Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1gs78wrxS34XhGPECmN5fABKpEuhUGwAXS"
  },
  {
    "address": "tz1iuZ76cLaaca8Dh4soYEHjYeVb5vHmmP5Z",
    "name": "Staking Facilities Payouts 2",
    "logo": "https://services.tzkt.io/v1/avatars/tz1iuZ76cLaaca8Dh4soYEHjYeVb5vHmmP5Z"
  },
  {
    "address": "tz1hxtCEcD7idQJEDiJEq37vBkoRcwF6KC2X",
    "name": "Staking Shop Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1hxtCEcD7idQJEDiJEq37vBkoRcwF6KC2X"
  },
  {
    "address": "tz1hTJnYMKSDUv55yREM8ezmC5nsdbPUWao7",
    "name": "Steak.and.Bake Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1hTJnYMKSDUv55yREM8ezmC5nsdbPUWao7"
  },
  {
    "address": "tz1iQRWn2zvdTqhJtGTBbYktodNtw67yE4a6",
    "name": "Tezbaguette Payouts ",
    "logo": "https://services.tzkt.io/v1/avatars/tz1iQRWn2zvdTqhJtGTBbYktodNtw67yE4a6"
  },
  {
    "address": "tz1TkQCxYhhpzMop7Da6g2tpa1Z2giUbPdqv",
    "name": "Tez Baker Payouts 5",
    "logo": "https://services.tzkt.io/v1/avatars/tz1TkQCxYhhpzMop7Da6g2tpa1Z2giUbPdqv"
  },
  {
    "address": "tz1g4o4a2wxpzJ7EgG3onFM5TLaPyiRFjFhL",
    "name": "Tezgate Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1g4o4a2wxpzJ7EgG3onFM5TLaPyiRFjFhL"
  },
  {
    "address": "tz1RiSGsJwcN6roErVt4D3mfSCYSaWYWP9q8",
    "name": "Tezoris Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1RiSGsJwcN6roErVt4D3mfSCYSaWYWP9q8"
  },
  {
    "address": "tz1LJycSuCT25AA5VJwNW1QYXVGyy7YLwZh9",
    "name": "TezosAirGap Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1LJycSuCT25AA5VJwNW1QYXVGyy7YLwZh9"
  },
  {
    "address": "tz1LXaBkg3gZLVyB2awuboPAi6CnWxT4bCMt",
    "name": "Tezos Boutique Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1LXaBkg3gZLVyB2awuboPAi6CnWxT4bCMt"
  },
  {
    "address": "tz1VfoA5qPksECBkvr2EUuT6u5VgLCi3vTPV",
    "name": "Tezos.nu Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1VfoA5qPksECBkvr2EUuT6u5VgLCi3vTPV"
  },
  {
    "address": "tz1Y2bfsQAYHrWZpM1AwomYMw2KEtaJ5VMfU",
    "name": "Tezos.Rio Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1Y2bfsQAYHrWZpM1AwomYMw2KEtaJ5VMfU"
  },
  {
    "address": "tz1LXz7C2wPr74nfMMHmqsqbnasPvi89o2U6",
    "name": "TezosRus Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1LXz7C2wPr74nfMMHmqsqbnasPvi89o2U6"
  },
  {
    "address": "tz1Tnr7CDL5i45EHSPpPReBocDiGbMwpdTCn",
    "name": "Tezos Wake n' Bake Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1Tnr7CDL5i45EHSPpPReBocDiGbMwpdTCn"
  },
  {
    "address": "tz1WntXgznbivRjyhE7Y5jEzoebAhMPB2iJa",
    "name": "Tezzieland Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1WntXgznbivRjyhE7Y5jEzoebAhMPB2iJa"
  },
  {
    "address": "tz1ant4etDpDJKQqASYri7bse5tCv2FNbvR6",
    "name": "XTZ Antipodes Payouts 2",
    "logo": "https://services.tzkt.io/v1/avatars/tz1ant4etDpDJKQqASYri7bse5tCv2FNbvR6"
  },
  {
    "address": "tz1Z7HGaayrL4XDNQ7UyXe7L4wXboELkhBm3",
    "name": "XTZ Delegate Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1Z7HGaayrL4XDNQ7UyXe7L4wXboELkhBm3"
  },
  {
    "address": "tz1aF8E89g1EqqzCmbNCLunTtCtxgkoQUX7N",
    "name": "0xb1 / PlusMinus Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1aF8E89g1EqqzCmbNCLunTtCtxgkoQUX7N"
  },
  {
    "address": "tz1KwdqMQGB75VDnSsLiteFKVX4HNzvbNvjZ",
    "name": "888XTZ Payouts",
    "logo": "https://services.tzkt.io/v1/avatars/tz1KwdqMQGB75VDnSsLiteFKVX4HNzvbNvjZ"
  },
  {
    "address": "tz1ea4wfxxvjpic1DZCU2WnUCBmiPVrSXrSz",
    "name": "Cerberus Bakery Payouts",
    "logo": null
  }
]
```
Could you add them as a separate JSON file and use here?

**Review by keshan3262 — DISMISSED** [review]

File: src/components/activity-groups-list/activity-group-item/activity-group-amount-change/activity-group-amount-change.tsx [hunk 9]
```diff
@@ -1,24 +1,39 @@
 import React, { FC } from 'react';
 import { Text, View } from 'react-native';
 
+import { shortizeSymbol } from 'src/utils/token-metadata.utils';
+
 import { NonZeroAmounts } from '../../../../interfaces/non-zero-amounts.interface';
 import { conditionalStyle } from '../../../../utils/conditional-style';
 import { formatAssetAmount } from '../../../../utils/number.util';
```
Comment (keshan3262) [inline src/components/activity-groups-list/activity-group-item/activity-group-amount-change/activity-group-amount-change.tsx:8]: Please write these imports with `src` as well

**Review by keshan3262 — APPROVED** [review]

**Review by keshan3262 — APPROVED** [review]

**Comment by herkoss** [comment]
> Please, don't ignore comments from previous reviewes

We have already discussed about your comments about bakers. Or what are you talking about?

File: src/utils/token-operations.util.ts [hunk 10]
```diff
@@ -1,85 +1,66 @@
...
+const getContractOperations = <T>(selectedRpcUrl: string, account: string, contractAddress: string) =>
   getTzktApi(selectedRpcUrl)
     .get<Array<T>>(`accounts/${contractAddress}/operations`, {
       params: {
         type: 'transaction',
         limit: OPERATION_LIMIT,
         sort: '1',
         initiator: account,
-        entrypoint: 'mintOrBurn',
-        ...(isDefined(lastLevel) ? { 'level.lt': lastLevel } : undefined)
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/utils/token-operations.util.ts:41]: Was fixed by id

File: src/utils/token-operations.util.ts [hunk 11]
```diff
@@ -1,85 +1,66 @@
...
   getTzktApi(selectedRpcUrl)
-    .get<Array<OperationFa12Interface>>('operations/transactions', {
+    .get<Array<Fa12TransferInterface>>('operations/transactions', {
       params: {
         limit: OPERATION_LIMIT,
         entrypoint: 'transfer',
         'sort.desc': 'level',
         target: contractAddress,
-        'parameter.in': `[{"from":"${account}"},{"to":"${account}"}]`,
-        ...(isDefined(lastLevel) ? { 'level.lt': lastLevel } : undefined)
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/utils/token-operations.util.ts:75]: Was fixed by id

**Review by keshan3262 — DISMISSED** [review]

File: src/hooks/use-contract-activity.ts [hunk 12]
```diff
@@ -1,36 +1,70 @@
...
-      const activities = await loadActivity(selectedRpcUrl, selectedAccount, tokenSlug);
+      const { activities, reachedTheEnd, oldestOperation } = await loadActivity(
+        selectedRpcUrl,
+        selectedAccount,
+        tokenSlug,
+        knownBakers
+      );
+      console.log('activities: ', activities);
+      console.log('reachedTheEnd: ', reachedTheEnd);
+      console.log('oldestOperation: ', oldestOperation);
```
Comment (keshan3262) [inline src/hooks/use-contract-activity.ts:54]: Please remove `console.log` statements

**Review by keshan3262 — DISMISSED** [review]

File: src/utils/token-operations.util.ts [hunk 13]
```diff
@@ -183,34 +188,50 @@ const loadOperations = async (
...
   selectedAccount: AccountInterface,
   tokenSlug?: string,
   knownBakers?: Array<TzktMemberInterface>,
-  lastItem?: Activity
-): Promise<Array<Array<Activity>>> => {
-  const operationsHashes = await loadOperations(selectedRpcUrl, selectedAccount, tokenSlug, lastItem)
-    .then(operations => operations.map(operation => operation.hash))
-    .then(newHashes => uniq(newHashes.filter(x => x !== lastItem?.hash)));
+  oldestOperation?: TzktOperation
+): Promise<{ activities: Array<Array<Activity>>; reachedTheEnd: boolean; oldestOperation?: TzktOperation }> => {
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/utils/token-operations.util.ts:205]: `Array<Activity>` is `ActivityGroup`, in this case `Array<Array<Activity>>` is Array<ActivityGroup>, this function returns the activity group

File: src/components/activity-groups-list/activity-groups-list.tsx [hunk 14]
```diff
@@ -19,84 +19,84 @@ interface Props {
...
+            onEndReached={() => {
+              if (!onEndReachedCalledDuringMomentum.current || sections.length < 10) {
+                handleUpdate();
+                onEndReachedCalledDuringMomentum.current = true;
+              }
+            }}
+            keyExtractor={item => item[0].hash}
+            bounces={false}
+            renderItem={({ item }) => renderItem(item)}
+            renderSectionHeader={({ section: { title } }) => <Text style={styles.sectionHeaderText}>{title}</Text>}
```
Comment (keshan3262) [inline src/components/activity-groups-list/activity-groups-list.tsx:95]: Please try wrapping callbacks here in `useCallback` hooks

File: src/hooks/use-contract-activity.ts [hunk 15]
```diff
@@ -16,6 +16,16 @@ import { loadActivity } from '../utils/token-operations.util';
 
 const OLDEST_ACTIVITY_INDEX = 0;
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/hooks/use-contract-activity.ts:17]: This is the index of oldest activity in group

File: src/hooks/use-contract-activity.ts [hunk 16]
```diff
@@ -43,31 +54,21 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
...
-          const allActivities = [...activities, ...prev];
-          const allHashes = allActivities.map(x => x[OLDEST_ACTIVITY_INDEX]).map(x => x.hash);
-          const onlyUniqueHashes = uniq(allHashes);
-          const onlyUniqueActivitites = onlyUniqueHashes
-            .map(x => allActivities.find(y => y[OLDEST_ACTIVITY_INDEX].hash === x))
-            .filter(isDefined);
-
-          return onlyUniqueActivitites;
-        });
+        setActivities(prev => getUniqActivities([...prev, ...newActivity]));
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/hooks/use-contract-activity.ts:69]: Because we got duplicates sometimes from TZKT api

File: src/components/activity-groups-list/activity-groups-list.tsx [hunk 17]
```diff
@@ -12,95 +12,126 @@ import { ActivityGroupItem } from './activity-group-item/activity-group-item';
...
-                      onImageError={onOptimalPromotionError}
-                      onEmptyPromotionReceived={onOptimalPromotionError}
-                    />
-                  </View>
-                )}
-              </>
-            )}
-            renderSectionHeader={({ section: { title } }) => <Text style={styles.sectionHeaderText}>{title}</Text>}
+            onEndReachedThreshold={0.5}
+            onMomentumScrollBegin={onMomentumScrollBegin}
```
[the remark this answers is not in the document]
Comment (herkoss) [reply src/components/activity-groups-list/activity-groups-list.tsx:125]: `onEndReached` calls many times in some cases and this is common problem, this fix is working

**Review by NovakVlad2424 — APPROVED** [review]
ios done 1.16_900

**Review by IrynaKhyzhynska — APPROVED** [review]
android done 1.16_900
